scrc/preprocessors/extractors/spider_specific/section_splitting_functions.py

```diff
@@ -24,6 +24,72 @@ def XX_SPIDER(decision: Union[bs4.BeautifulSoup, str], namespace: dict) -> Optio
+    """
+    # As soon as one of the strings in the list (regexes) is encountered we switch to the corresponding section (key)
+    # (?:C|c) is much faster for case insensitivity than [Cc] or (?i)c
+    all_section_markers = {
+        Language.DE: {
+            Section.FACTS: [r'^Sachverhalt:?\s*$', r'^Tatsachen$'],
+            Section.CONSIDERATIONS: [r'^Begründung:\s*$',r'Erwägung(en)?:?\s*$',r'^Entscheidungsgründe$', r'[iI]n Erwägung[:,]?\s*$'],
+            Section.RULINGS: [r'Demgemäss erkennt d[\w]{2}', r'erkennt d[\w]{2} [A-Z]\w+:', r'Appellationsgericht (\w+ )?(\(\w+\) )?erkennt', r'^und erkennt:$', r'erkennt:\s*$'],
+            Section.FOOTER: [r'^Rechtsmittelbelehrung$',
+                             r'AUFSICHTSKOMMISSION', r'APPELLATIONSGERICHT']
+        }
+    }
+    if namespace['language'] not in all_section_markers:
+        message = f"This function is only implemented for the languages {list(all_section_markers.keys())} so far."
+        raise ValueError(message)
+
+    section_markers = all_section_markers[namespace['language']]
+
+    # combine multiple regex into one for each section due to performance reasons
+    section_markers = dict(
+        map(lambda kv: (kv[0], '|'.join(kv[1])), section_markers.items()))
+
+    # normalize strings to avoid problems with umlauts
+    for section, regexes in section_markers.items():
+        section_markers[section] = unicodedata.normalize('NFC', regexes)
+        # section_markers[key] = clean_text(regexes) # maybe this would solve some problems because of more cleaning
+
+    def get_paragraphs(soup):
+        """
+        Get Paragraphs in the decision
+        :param soup:
+        :return:
+        """
+
+        divs = soup.find_all(
+            "div", class_=['WordSection1', 'Section1', 'WordSection2'])
+        paragraphs = []
+        heading, paragraph = None, None
+        for el in divs:
+            for element in el:
+                if isinstance(element, bs4.element.Tag):
+                    text = str(element.string)
+                    # This is a hack to also get tags which contain other tags such as links to BGEs
+                    if text.strip() == 'None':
+                        text = element.get_text()
+                    # get numerated titles such as 1. or A.
+                    if "." in text and len(text) < 5:
+                        heading = text  # set heading for the next paragraph
+                    else:
+                        if heading is not None:  # if we have a heading
+                            paragraph = heading + " " + text  # add heading to text of the next paragraph
+                        else:
+                            paragraph = text
+                        heading = None  # reset heading
+                    paragraph = clean_text(paragraph)
+                    if paragraph not in ['', ' ', None]:  # discard empty paragraphs
+                        paragraphs.append(paragraph)
```

Comment:
Dies ist dasselbe wie beim BGer. Da dieses File rasch sehr gross werden kann ist die Vermeidung von Code Duplikation wichtig. Versuche also so viele funktionen wie möglich zu extrahieren, welche in mehreren Spiders gebraucht werden (in deinem Fall BS_Omni und CH_BGer).

Reply:
Ich habe im neusten Commit 3 Funktionen extracted, welche jeweils Funktionalitäten beinhalteten, die schon bei CH_BGer verwendet wurden um Code Duplikation zu verhindern: 

- get_paragraphs
- valid_name
- prepare_section_markers

Dabei musste ich die get_paragraphs implementation auf 2 Schleifen verallgemeinern, da bei CH_BGer davon ausgegangen wurde, dass der gesamte Content in `div[0]` gefunden werden kann. Dabei können für andere Gerichte mehrere divs mit Content vorkommen.

Reply:
Sehr gut